### PR TITLE
Add ExtraArg merging and RFS Extra Arg Description

### PR DESCRIPTION
**Pull request by AndreKurait**
### Description
Adds description of reindexFromSnapshotExtraArgs to options.md
Adds support on ExtraArgs parameters for overriding of existing parameters and boolean negation (--no prefix removes flag)
Adds RFS stack tests

* Category: New Feature/Documentation
* Why these changes are required? Allow users to override RFS parameters like with a custom snapshot
* What is the old behavior before changes and new behavior after changes? No Documentation on RFS Extra args before, no support for merging args

### Issues Resolved
[MIGRATIONS-1908](https://opensearch.atlassian.net/browse/MIGRATIONS-1908)

Is this a backport? If so, please add backport PR # and/or commits #

### Testing
Thorough unit testing

### Check List
- [x] New functionality includes testing
  - [x] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: deployment/cdk/opensearch-service-migration/lib/common-utilities.ts [hunk 1]
```diff
@@ -6,6 +6,78 @@ import { IApplicationLoadBalancer } from "aws-cdk-lib/aws-elasticloadbalancingv2
...
+    // Reconstruct command
+    const mergedArgs = Object.entries(mergedArgv)
+        .filter(([key]) => key !== '_' && key !== '$0')
+        .map(([key, value]) => {
+            if (typeof value === 'boolean') {
+                return value ? `--${key}` : `--no-${key}`;
+            }
+            return `--${key} ${value}`;
+        })
+        .join(' ');
```
Comment (peternied) [inline deployment/cdk/opensearch-service-migration/lib/common-utilities.ts:75]: Is this list still in the order of the original command structure with the additional args in order at the end?  If not can we make it work that way to make diffing easier?

File: deployment/cdk/opensearch-service-migration/lib/common-utilities.ts [hunk 2]
```diff
@@ -6,6 +6,78 @@ import { IApplicationLoadBalancer } from "aws-cdk-lib/aws-elasticloadbalancingv2
...
+
+    let fullCommand = `${commandPrefix} ${mergedArgs}`.trim()
+    return fullCommand;
```
Comment (peternied) [inline deployment/cdk/opensearch-service-migration/lib/common-utilities.ts:78]: Can we log the baseCommand and fullCommand?

File: deployment/cdk/opensearch-service-migration/options.md [hunk 3]
```diff
@@ -39,10 +39,31 @@ These tables list all CDK context configuration values a user can specify for th
...
+|-----------------------------------|---------|----------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------|
+| reindexFromSnapshotServiceEnabled | boolean | true                                                                 | Create resources for deploying and configuring the RFS ECS service                                                                       |
+| reindexFromSnapshotExtraArgs      | string  | "--target-aws-region us-east-1 --target-aws-service-signing-name es" | Extra arguments to provide to the Document Migration command with space separation. See RFS Extra Arguments below for available options. Boolean negation is used, so --no-insecure can be passed to remove the --insecure flag if present. |
+| sourceClusterEndpoint             | string  | `"https://source-cluster.elb.us-east-1.endpoint.com"`                | The endpoint for the source cluster from which RFS will take a snapshot                                                                  |
+
+#### RFS Extra Arguments
```
Comment (mikaylathompson) [inline deployment/cdk/opensearch-service-migration/options.md:48]: I'm not sure I would actually document all of this here. Do we already have it documented in an RFS readme? Can we just point to it there? (and if not, can we move this table there?)

We expose extraArgs for a lot of tools and I don't want us to have to maintain documentation for all possible args in multiple places. Additionally, I think we're slightly more likely to confuse people by making them think they have to specify e.g. a snapshot name, when usually we handle it.

File: deployment/cdk/opensearch-service-migration/lib/common-utilities.ts (hunk 1 above)
Comment (peternied) [reply deployment/cdk/opensearch-service-migration/lib/common-utilities.ts:75]: It looks like this is the case, but I'm double checking

**Review by mikaylathompson — APPROVED** [review]
Documentation comments, can be addressed in a followup

File: deployment/cdk/opensearch-service-migration/options.md [hunk 4]
```diff
@@ -39,10 +39,31 @@ These tables list all CDK context configuration values a user can specify for th
...
+| Name                              | Type    | Example                                                              | Description                                                                                                                              |
+|-----------------------------------|---------|----------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------|
+| reindexFromSnapshotServiceEnabled | boolean | true                                                                 | Create resources for deploying and configuring the RFS ECS service                                                                       |
+| reindexFromSnapshotExtraArgs      | string  | "--target-aws-region us-east-1 --target-aws-service-signing-name es" | Extra arguments to provide to the Document Migration command with space separation. See RFS Extra Arguments below for available options. Boolean negation is used, so --no-insecure can be passed to remove the --insecure flag if present. |
```
Comment (peternied) [inline deployment/cdk/opensearch-service-migration/options.md:45]: Can you add a footnote with with example of adding, overriding, and removing - then reference it in all the extra arguments for snapshot, metadata migration, rfs, capture proxy, replayer?

How to Footnote; https://github.blog/changelog/2021-09-30-footnotes-now-supported-in-markdown-fields/

**Review by peternied — APPROVED** [review]
Fancy!  This looks like it will do well to get folks unblocked

File: deployment/cdk/opensearch-service-migration/options.md (hunk 4 above)
Comment (AndreKurait) [reply deployment/cdk/opensearch-service-migration/options.md:45]: Will address in followup PR due to github actions flakiness

File: deployment/cdk/opensearch-service-migration/lib/common-utilities.ts (hunk 2 above)
Comment (AndreKurait) [reply deployment/cdk/opensearch-service-migration/lib/common-utilities.ts:78]: Will address in followup PR due to github actions flakiness

File: deployment/cdk/opensearch-service-migration/options.md (hunk 3 above)
Comment (AndreKurait) [reply deployment/cdk/opensearch-service-migration/options.md:48]: Will address in followup PR due to github actions flakiness

File: deployment/cdk/opensearch-service-migration/options.md (hunk 4 above)
Comment (AndreKurait) [reply deployment/cdk/opensearch-service-migration/options.md:45]: https://github.com/opensearch-project/opensearch-migrations/pull/864

File: deployment/cdk/opensearch-service-migration/lib/common-utilities.ts (hunk 2 above)
Comment (AndreKurait) [reply deployment/cdk/opensearch-service-migration/lib/common-utilities.ts:78]: Commands here may have sensitive information like plaintext password, i'll skip this print statement

File: deployment/cdk/opensearch-service-migration/options.md (hunk 3 above)
Comment (AndreKurait) [reply deployment/cdk/opensearch-service-migration/options.md:48]: https://github.com/opensearch-project/opensearch-migrations/pull/864

**Pull request merged**